File: scala-train/src/main/scala/com/tanknavy/scala_package/packageObject.scala
package com.tanknavy.scala_package

/**
 * Author: Alex Cheng 9/15/2020 7:38 PM
 */

//包对象
//package com.tanknavy.scala_package {
package bag {

  object packageTest {
    def main(args: Array[String]): Unit = {
      println("packageTest")
      thePack.add() //同一文件夹下的包对象thePack

      import com.tanknavy.scala_package.myPack._ //另外一个文件件下
      val cus = new Customer()
      println(cus.id)
      println(cus.name)
      sayHi()

    }
  }

  //包对象
  //1.在包中可以包含pacakge/class/object/trait, 但是不能包含variable, function，这是jvm的局限 ==>使用包对象的技术
  //2.package object myPackage表示创建一个包对象，必须同名
  //3.每一个包都可以有一个包对象，切只有一个,类似Python的包中__init__文件中可以写一些变量，方法，其它模块都可以引用!!!
  //4.在包对象中定义的变量和方法，就可以在对应的包中使用
  package object thePack {
    var objectName: String = "bob"

    def add(): Unit = {
      println("package object thePack")
    }
  }

  package thePack {

    class Student {
      var name: String = ""
      var id: Int = 0
      println("Student class" + objectName)
    }

  }


}

File: scala-train/src/main/java/com/tanknavy/bag/PackageObject.java
package com.tanknavy.bag;

/**
 * Author: Alex Cheng 9/15/2020 7:50 PM
 */
//模拟Scala中的包对象，

public class PackageObject {
    public static void main(String[] args) {
        System.out.println(Package$$.class.getName());
        System.out.println("--------------");

        System.out.println(Package$$.module$.name());//通过类的静态属性
        Package$$.module$.sayHi();

        System.out.println("--------------");
        Package$$ sp = new Package$$();
        System.out.println(sp.name());
        sp.sayHi();
    }
}

//模拟scala的包对象(包含一个name属性和一个方法)，关键是static module
//scala类编译时会产生两个文件，
//class中静态属性module, 然后这个属性使用今天
final class Package$${
    //定义属性，包括今天对象module$
    public static final Package$$ module$; //怎么初始化?
    //public static final Package$$ module$ = new Package$$(); //怎么初始化?

    private String name = "king"; //模拟scala包对象中的属性

    public Package$$() {
        System.out.println("blank constructor...");
    }

    static { //静态代码块调用构造方法完成静态对象module$的初始化！！
        System.out.println("static start...");
        module$ = new Package$$();
        System.out.println("static end...");
    };

    public String name(){ //属性的get方法
        return this.name;
    }
    public void sayHi(){ //模拟scala包对象中的方法
        System.out.println("scala package object method...");
    }
}
